common/remote_util.cc

```diff
@@ -158,33 +144,61 @@ ProcessStartInfo RemoteUtil::BuildProcessStartInfoForSshInternal(
       "-oServerAliveCountMax=6 "  // Number of lost msgs before ssh terminates
       "-oServerAliveInterval=5 "  // Time interval between alive msgs
       "%s %s -p %i %s",
-      ssh_command_, quiet_ || verbosity_ < 2 ? "-q" : "", forward_arg,
-      QuoteArgument(user_host_), ssh_port_, remote_command_arg);
+      Quote(ssh_command_), quiet_ || verbosity_ < 2 ? "-q" : "", forward_arg,
+      Quote(user_host_), ssh_port_, remote_command_arg);
   start_info.forward_output_to_log = forward_output_to_log_;
   return start_info;
 }
 
-std::string RemoteUtil::EscapeForWindows(const std::string& argument) {
-  std::string str =
+std::string RemoteUtil::Quote(const std::string& argument) {
+  // Escape certain backslashes (see doc of this function).
+  std::string escaped =
       std::regex_replace(argument, std::regex(R"(\\*(?="|$))"), "$&$&");
-  return std::regex_replace(str, std::regex(R"(")"), R"(\")");
-}
+  // Escape " -> \".
+  escaped = std::regex_replace(escaped, std::regex(R"(")"), R"(\")");
+  // Quote.
+  return absl::StrCat("\"", escaped, "\"");
+}
+
+std::string RemoteUtil::QuoteForSsh(const std::string& argument) {
+  // Escape \ ->: \\.
+  std::string escaped =
+      std::regex_replace(argument, std::regex(R"(\\)"), R"(\\)");
+  // Escape " -> \".
+  escaped = std::regex_replace(escaped, std::regex(R"(")"), R"(\")");
+
+  // Quote, but handle special case for ~.
+  if (escaped.empty() || escaped[0] != '~') {
+    return Quote(absl::StrCat("\"", escaped, "\""));
+  }
 
-std::string RemoteUtil::QuoteArgument(const std::string& argument) {
-  return absl::StrCat("\"", EscapeForWindows(argument), "\"");
-}
+  // Simple special cases. Quote() isn't required, but called for consistency.
+  if (escaped == "~" || escaped == "~/") {
+    return Quote(escaped);
```

Comment:
Doesn't that wrap `~` in quotes and break the expansion?

Reply:
Yes, but only in the outer quotes that get removed by Windows. It won't get wrapped into the additional inner quotes that get removed by Linux. Those would break the expansion.